bluepyopt/ephys/create_acc.py

```diff
@@ -0,0 +1,832 @@
+
+
+class Nrn2ArbMechGrouper:
+    """Group parameters by mechanism and convert them to Arbor format"""
+
+    @staticmethod
+    def _is_global_property(loc, param):
+        """Returns if a label-specific variable is a global property in Arbor
+
+        Args:
+            loc (): An Arbor label describing the location
+            param (): A parameter in Arbor format (name and units)
+        """
+
+        return loc == ArbFileMorphology.region_labels['all'] and (
+            param.name in ['membrane-potential',
+                           'temperature-kelvin',
+                           'axial-resistivity',
+                           'membrane-capacitance'] or
+            param.name.split(' ')[0] in ['ion-internal-concentration',
+                                         'ion-external-concentration',
+                                         'ion-reversal-potential'])
+
+    @classmethod
+    def _separate_global_properties(cls, loc, mechs):
+        """Separates global properties from a label-specific dict of mechanisms
+
+        Args:
+            loc (): An Arbor label describing the location
+            mechs (): A mapping of mechanism name to list of parameters in
+            Arbor format (None for non-mechanism parameters).
+
+        Returns:
+            A split of mechs into mechanisms without Arbor global properties
+            (first component) and a dict with Arbor global properties
+            (second component)
+        """
+
+        local_mechs = dict()
+        global_properties = []
+
+        for mech, params in mechs.items():
+
+            if mech is None:
+                local_properties = []
+                for param in params:
+                    if cls._is_global_property(loc, param):
+                        global_properties.append(param)
+                    else:
+                        local_properties.append(param)
+                local_mechs[mech] = local_properties
+            else:
+                local_mechs[mech] = params
+
+        return local_mechs, {None: global_properties}
+
+    @staticmethod
+    def _format_params_and_group_by_mech(params, channels):
+        """Group list of parameters by mechanism and turn them to Arbor format
+
+        Args:
+            params (): List of parameters in Neuron format
+            channels (): List of co-located NMODL mechanisms
+
+        Returns:
+            Mapping of Arbor mechanism name to list of parameters in Arbor
+            format
+        """
+        mech_params = [Nrn2ArbParamFormatter.format(
+                       param, channels) for param in params]
+        mechs = {mech: [] for mech, _ in mech_params}
+        for mech in channels:
+            if mech not in mechs:
+                mechs[mech] = []
+        for mech, param in mech_params:
+            mechs[mech].append(param)
+        return mechs
+
+    @classmethod
+    def process_global(cls, params):
```

Comment:
Neither of the latter is needed really. Both methods do what they say - grouping _BluePyOpt_ parameters by mechanism and converting them to Arbor format. The difference is in the input argument, not the return values. If they're difficult to understand, looking at their invocation context may help.

Reply:
Ah sorry my previous comment above was for here. `process_local` returns a tuple. Could you mention it in the docstring?

Reply:
My confusion is because `process_local` also returns the global ones.
Since there already exists a `process_global`, I expect `process_local` to only return the local ones but it is also returning the global ones.

I don't understand how the global ones returned from `process_local` are different than the global ones returned from `process_global`. I expect the function names to explain this.

Reply:
The docstring is fixed in 04ce61439f8942250b71f7b9f684d6040be14773. For the confusion with `process_global`/`process_local`, it's important to keep the name of its encapsulating class in mind: `Nrn2ArbMechGrouper`. These two methods define the interface of that class and, thus, map parameters that come in Neuron format as extracted in the first step (shared between `create_hoc` and `create_acc`) to Arbor format. The `_local` and `_global` qualifiers refer to the input arguments, thus, Neuron's representation. The output, though, is Arbor's view. The name global property is reserved in Arbor and the way to represent this in a decor is with the `default` keyword. As you can see from the invocation context, global properties obtained derived from `global_params`, `section_params` and `range_params` all get absorbed into what eventually becomes `global_mechs` and `global_scaled_mechs` in the Jinja templates.